#RESTART:
- new dataset from kaggle https://www.kaggle.com/datasets/paramaggarwal/fashion-product-images-dataset?resource=download

##Mounting Drive

In [1]:
import subprocess
subprocess.run(['umount', '-l', '/content/drive'], capture_output=True)

from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [2]:
# ===================================
# CELL 1: DOWNLOAD (SKIP IF EXISTS)
# ===================================

from google.colab import drive
import pandas as pd
import shutil
import os
import json
import time

drive.mount('/content/drive')

!pip install -q kaggle

os.makedirs('/root/.kaggle', exist_ok=True)

kaggle_credentials = {
    "username": "biancagambino",
    "key": "4cfe6253c2484adad8e890df33cdf855"
}

with open('/root/.kaggle/kaggle.json', 'w') as f:
    json.dump(kaggle_credentials, f)

!chmod 600 /root/.kaggle/kaggle.json

print("✅ Kaggle credentials set up")

# Check if already downloaded
if os.path.exists('fashion-product-images-dataset.zip'):
    file_size = os.path.getsize('fashion-product-images-dataset.zip')
    print(f"\n✅ Zip already exists ({file_size / (1024*1024):.1f} MB)")
    print("Skipping download...")
else:
    print("\n📥 Downloading dataset...")
    !wget -O fashion-product-images-dataset.zip https://www.kaggle.com/api/v1/datasets/download/paramaggarwal/fashion-product-images-dataset

    if os.path.exists('fashion-product-images-dataset.zip'):
        print("✅ Download complete!")
    else:
        print("❌ Download failed")

print("\n✅ Ready for extraction (run Cell 2)")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
✅ Kaggle credentials set up

📥 Downloading dataset...
--2026-03-26 21:32:16--  https://www.kaggle.com/api/v1/datasets/download/paramaggarwal/fashion-product-images-dataset
Resolving www.kaggle.com (www.kaggle.com)... 35.244.233.98
Connecting to www.kaggle.com (www.kaggle.com)|35.244.233.98|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://storage.googleapis.com:443/kaggle-data-sets/139630/329006/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20260326%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20260326T213217Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=81967dcf46e03e7521c8e7d3745b94ab34d7300a3e9f19a97f1977343e4154f2ce4288f164f17b9f223195aa202c070461af973bfe318b9fc29bab0a6f905c91a78fde1b7a82fc0569c32bf74579dbe94f1d5aa4fa8

##Extract/Copy and Delete duplicated zip files

In [3]:
# ===================================
# CELL 2: EXTRACT AND COPY - FIXED
# ===================================

import pandas as pd
import shutil
import os
import glob

# Remove old extracted files
print("🧹 Removing old extracted files...")
!rm -rf /content/extracted

print("📦 Extracting dataset (5-10 minutes)...")
!unzip -o -q fashion-product-images-dataset.zip -d /content/extracted
print("✅ Extract complete!")

# ===================================
# FIND FILES IN /content/extracted ONLY
# ===================================
print("\n🔍 Finding dataset files...")

csv_files = glob.glob('/content/extracted/**/styles.csv', recursive=True)
print(f"Found CSVs: {csv_files}")

if not csv_files:
    print("❌ styles.csv not found! Here's what was extracted:")
    !find /content/extracted -type f -name "*.csv"
    !find /content/extracted -type d
    raise FileNotFoundError("Could not find styles.csv — check output above")

styles_path = csv_files[0]
dataset_dir = os.path.dirname(styles_path)
images_dir = os.path.join(dataset_dir, 'images')

# Fallback if images folder is named differently
if not os.path.exists(images_dir):
    image_dirs = glob.glob('/content/extracted/**/images', recursive=True)
    if image_dirs:
        images_dir = image_dirs[0]
    else:
        print("❌ images folder not found! Here's what exists:")
        !find /content/extracted -type d
        raise FileNotFoundError("Could not find images folder")

print(f"✅ styles.csv: {styles_path}")
print(f"✅ images dir: {images_dir}")

num_images = len([f for f in os.listdir(images_dir) if f.endswith('.jpg')])
print(f"   Total images found: {num_images}")

if num_images < 1000:
    print("⚠️  WARNING: Less than 1000 images found — something may be wrong")

# ===================================
# LOAD AND FILTER CATEGORIES
# ===================================
print("\n📊 Loading CSV...")
df = pd.read_csv(styles_path, on_bad_lines='skip')
print(f"Total rows: {len(df)}")

# Only keep the categories useful for a clothing app
KEEP_CATEGORIES = [
    'Tshirts', 'Shirts', 'Jeans', 'Dresses', 'Tops',
    'Trousers', 'Shorts', 'Jackets', 'Sweaters',
    'Kurtas', 'Casual Shoes', 'Formal Shoes', 'Sandals',
    'Handbags', 'Watches'
]

df = df[df['articleType'].isin(KEEP_CATEGORIES)]
print(f"Rows after category filter: {len(df)}")
print(f"\nCategory counts:")
print(df['articleType'].value_counts().to_string())

# Sample up to 500 per category (balanced dataset)
df_sampled = df.groupby('articleType', group_keys=False).apply(
    lambda x: x.sample(min(len(x), 500), random_state=42)
)

print(f"\nFinal sample size: {len(df_sampled)} images")

# ===================================
# COPY TO GOOGLE DRIVE
# ===================================
print("\n📤 Copying to Google Drive...")
base_path = '/content/drive/MyDrive/ClosetGeniusCNN'
new_images_path = f'{base_path}/images'

if os.path.exists(new_images_path):
    shutil.rmtree(new_images_path)
os.makedirs(new_images_path, exist_ok=True)

copied = 0
missing = 0
for idx, row in df_sampled.iterrows():
    src = os.path.join(images_dir, f"{row['id']}.jpg")
    dst = os.path.join(new_images_path, f"{row['id']}.jpg")
    if os.path.exists(src):
        shutil.copy(src, dst)
        copied += 1
        if copied % 500 == 0:
            print(f"  ✓ {copied} images copied...")
    else:
        missing += 1

# Save filtered CSV
df_sampled.to_csv(f'{base_path}/styles.csv', index=False)

images_csv = os.path.join(dataset_dir, 'images.csv')
if os.path.exists(images_csv):
    shutil.copy(images_csv, f'{base_path}/images.csv')

print(f"\n🎉 DONE!")
print(f"✅ Copied: {copied} images")
print(f"⚠️  Missing: {missing} images (not in zip)")
print(f"📁 Saved to: {base_path}")
print(f"\n✅ READY TO TRAIN!")

🧹 Removing old extracted files...
📦 Extracting dataset (5-10 minutes)...
✅ Extract complete!

🔍 Finding dataset files...
Found CSVs: ['/content/extracted/fashion-dataset/styles.csv', '/content/extracted/fashion-dataset/fashion-dataset/styles.csv']
✅ styles.csv: /content/extracted/fashion-dataset/styles.csv
✅ images dir: /content/extracted/fashion-dataset/images
   Total images found: 44441

📊 Loading CSV...
Total rows: 44424
Rows after category filter: 25255

Category counts:
articleType
Tshirts         7067
Shirts          3217
Casual Shoes    2845
Watches         2542
Kurtas          1844
Tops            1762
Handbags        1759
Sandals          897
Formal Shoes     637
Jeans            609
Shorts           547
Trousers         530
Dresses          464
Sweaters         277
Jackets          258

Final sample size: 6999 images

📤 Copying to Google Drive...


/tmp/ipykernel_30438/3162748468.py:76: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df_sampled = df.groupby('articleType', group_keys=False).apply(


  ✓ 500 images copied...
  ✓ 1000 images copied...
  ✓ 1500 images copied...
  ✓ 2000 images copied...
  ✓ 2500 images copied...
  ✓ 3000 images copied...
  ✓ 3500 images copied...
  ✓ 4000 images copied...
  ✓ 4500 images copied...
  ✓ 5000 images copied...
  ✓ 5500 images copied...
  ✓ 6000 images copied...
  ✓ 6500 images copied...

🎉 DONE!
✅ Copied: 6998 images
⚠️  Missing: 1 images (not in zip)
📁 Saved to: /content/drive/MyDrive/ClosetGeniusCNN

✅ READY TO TRAIN!


In [4]:
import os

# Delete the zip (already extracted, safe to remove)
!rm -f /content/fashion-product-images-dataset.zip

# Delete the extracted folder too
!rm -rf /content/extracted

# Check how much space you freed
!df -h /content

Filesystem      Size  Used Avail Use% Mounted on
overlay         236G   46G  190G  20% /


##Full model Run

In [5]:
#MAIN DEFININTIONS
import torch
import torch.nn as nn
from torchvision import models

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
DATASET_PATH = "/content/drive/MyDrive/ClosetGeniusCNN/"

class MultiOutputAlexNet(nn.Module):
    def __init__(self, num_categories, num_colors, num_seasons, num_usage):
        super(MultiOutputAlexNet, self).__init__()
        alexnet = models.alexnet(weights=models.AlexNet_Weights.IMAGENET1K_V1)
        self.features = alexnet.features
        self.avgpool  = alexnet.avgpool
        for i, layer in enumerate(self.features):
            if i < 8:
                for param in layer.parameters():
                    param.requires_grad = False
        self.shared = nn.Sequential(
            nn.Dropout(0.5),
            nn.Linear(256 * 6 * 6, 4096),
            nn.ReLU(inplace=True),
            nn.Dropout(0.5),
            nn.Linear(4096, 2048),
            nn.ReLU(inplace=True)
        )
        self.category_head = nn.Linear(2048, num_categories)
        self.color_head    = nn.Linear(2048, num_colors)
        self.season_head   = nn.Linear(2048, num_seasons)
        self.usage_head    = nn.Linear(2048, num_usage)

    def forward(self, x):
        x = self.features(x)
        x = self.avgpool(x)
        x = torch.flatten(x, 1)
        x = self.shared(x)
        return (
            self.category_head(x),
            self.color_head(x),
            self.season_head(x),
            self.usage_head(x)
        )

print(f"Device: {device}")
print("MultiOutputAlexNet defined.")

Device: cuda:0
MultiOutputAlexNet defined.


###Actual Model Training

In [ ]:
# ACTUAL MODEL TRAINING
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from torchvision import transforms
from PIL import Image
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from datetime import datetime
import json

SAVE_PATH  = os.path.join(DATASET_PATH, "models")
LOGS_PATH  = os.path.join(DATASET_PATH, "logs")
PLOTS_PATH = os.path.join(DATASET_PATH, "plots")

os.makedirs(SAVE_PATH, exist_ok=True)
os.makedirs(LOGS_PATH, exist_ok=True)
os.makedirs(PLOTS_PATH, exist_ok=True)

EXPERIMENTS = {
    'experiment_1': {'name': 'Frozen_Baseline',  'batch_size': 64, 'learning_rate': 0.001, 'augmentation': 'basic'},
    'experiment_2': {'name': 'Frozen_High_LR',   'batch_size': 64, 'learning_rate': 0.005, 'augmentation': 'basic'}
}

df = pd.read_csv(os.path.join(DATASET_PATH, "styles.csv"), on_bad_lines='skip')
df['image_path'] = df['id'].apply(lambda x: os.path.join(DATASET_PATH, "images", f"{x}.jpg"))
df['exists']     = df['image_path'].apply(os.path.isfile)
df = df[df['exists']].reset_index(drop=True)
df = df.dropna(subset=['articleType', 'baseColour', 'season', 'usage'])

le_category = LabelEncoder()
le_color    = LabelEncoder()
le_season   = LabelEncoder()
le_usage    = LabelEncoder()

df['category_label'] = le_category.fit_transform(df['articleType'])
df['color_label']    = le_color.fit_transform(df['baseColour'])
df['season_label']   = le_season.fit_transform(df['season'])
df['usage_label']    = le_usage.fit_transform(df['usage'])

train_df, val_df = train_test_split(df, test_size=0.2, random_state=42, stratify=df['category_label'])
print(f"Train: {len(train_df)} | Val: {len(val_df)}")

class FashionDataset(Dataset):
    def __init__(self, dataframe, transform=None):
        self.df        = dataframe.reset_index(drop=True)
        self.transform = transform
    def __len__(self): return len(self.df)
    def __getitem__(self, idx):
        try:
            image = Image.open(self.df.iloc[idx]['image_path']).convert('RGB')
        except:
            image = Image.new('RGB', (224, 224), (0, 0, 0))
        if self.transform: image = self.transform(image)
        return (image, self.df.iloc[idx]['category_label'], self.df.iloc[idx]['color_label'],
                self.df.iloc[idx]['season_label'], self.df.iloc[idx]['usage_label'])

def get_transforms(augmentation_type='basic'):
    if augmentation_type == 'basic':
        train_transform = transforms.Compose([
            transforms.Resize((224, 224)), transforms.RandomHorizontalFlip(p=0.5),
            transforms.ToTensor(), transforms.Normalize([0.485,0.456,0.406],[0.229,0.224,0.225])
        ])
    else:
        train_transform = transforms.Compose([
            transforms.Resize((256,256)), transforms.RandomCrop((224,224)),
            transforms.RandomHorizontalFlip(p=0.5), transforms.RandomRotation(30),
            transforms.ColorJitter(brightness=0.5, contrast=0.5, saturation=0.5, hue=0.2),
            transforms.RandomGrayscale(p=0.1),
            transforms.RandomApply([transforms.GaussianBlur(kernel_size=5, sigma=(0.1,2.0))], p=0.3),
            transforms.RandomPerspective(distortion_scale=0.4, p=0.4),
            transforms.RandomResizedCrop(224, scale=(0.6,1.0)), transforms.ToTensor(),
            transforms.RandomErasing(p=0.2, scale=(0.02,0.15)),
            transforms.Normalize([0.485,0.456,0.406],[0.229,0.224,0.225])
        ])
    val_transform = transforms.Compose([
        transforms.Resize((224,224)), transforms.ToTensor(),
        transforms.Normalize([0.485,0.456,0.406],[0.229,0.224,0.225])
    ])
    return train_transform, val_transform

def train_one_epoch(model, dataloader, criterion, optimizer, device):
    model.train()
    running_loss, total = 0.0, 0
    correct = {'category':0,'color':0,'season':0,'usage':0}
    for images, categories, colors, seasons, usages in dataloader:
        images, categories = images.to(device), categories.to(device)
        colors, seasons, usages = colors.to(device), seasons.to(device), usages.to(device)
        optimizer.zero_grad()
        cat_out, col_out, sea_out, usa_out = model(images)
        loss = (criterion(cat_out,categories) + 0.8*criterion(col_out,colors) +
                0.6*criterion(sea_out,seasons) + 0.7*criterion(usa_out,usages))
        loss.backward(); optimizer.step()
        running_loss += loss.item(); total += categories.size(0)
        correct['category'] += (cat_out.argmax(1)==categories).sum().item()
        correct['color']    += (col_out.argmax(1)==colors).sum().item()
        correct['season']   += (sea_out.argmax(1)==seasons).sum().item()
        correct['usage']    += (usa_out.argmax(1)==usages).sum().item()
    return running_loss/len(dataloader), {k: 100*v/total for k,v in correct.items()} | {'average': sum(100*v/total for v in
correct.values())/4}

def validate(model, dataloader, criterion, device):
    model.eval()
    running_loss, total = 0.0, 0
    correct = {'category':0,'color':0,'season':0,'usage':0}
    with torch.no_grad():
        for images, categories, colors, seasons, usages in dataloader:
            images, categories = images.to(device), categories.to(device)
            colors, seasons, usages = colors.to(device), seasons.to(device), usages.to(device)
            cat_out, col_out, sea_out, usa_out = model(images)
            loss = (criterion(cat_out,categories) + 0.8*criterion(col_out,colors) +
                    0.6*criterion(sea_out,seasons) + 0.7*criterion(usa_out,usages))
            running_loss += loss.item(); total += categories.size(0)
            correct['category'] += (cat_out.argmax(1)==categories).sum().item()
            correct['color']    += (col_out.argmax(1)==colors).sum().item()
            correct['season']   += (sea_out.argmax(1)==seasons).sum().item()
            correct['usage']    += (usa_out.argmax(1)==usages).sum().item()
    return running_loss/len(dataloader), {k: 100*v/total for k,v in correct.items()} | {'average': sum(100*v/total for v in
correct.values())/4}

def run_experiment(experiment_config, train_df, val_df, num_epochs=15):
    exp_name, batch_size = experiment_config['name'], experiment_config['batch_size']
    learning_rate, augmentation = experiment_config['learning_rate'], experiment_config['augmentation']
    print(f"\nEXPERIMENT: {exp_name} | Batch:{batch_size} LR:{learning_rate} Aug:{augmentation}")
    train_transform, val_transform = get_transforms(augmentation)
    train_loader = DataLoader(FashionDataset(train_df, train_transform), batch_size=batch_size, shuffle=True,  num_workers=2,
pin_memory=True)
    val_loader   = DataLoader(FashionDataset(val_df,   val_transform),   batch_size=batch_size, shuffle=False, num_workers=2,
pin_memory=True)
    model = MultiOutputAlexNet(len(le_category.classes_), len(le_color.classes_), len(le_season.classes_),
len(le_usage.classes_)).to(device)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
    scheduler = optim.lr_scheduler.StepLR(optimizer, step_size=5, gamma=0.1)
    best_val_acc, model_save_path = 0.0, os.path.join(SAVE_PATH, f'{exp_name}_best.pth')
    for epoch in range(num_epochs):
        train_loss, train_acc = train_one_epoch(model, train_loader, criterion, optimizer, device)
        val_loss,   val_acc   = validate(model, val_loader, criterion, device)
        scheduler.step()
        print(f"Epoch [{epoch+1}/{num_epochs}] Train Avg:{train_acc['average']:.1f}% Val Avg:{val_acc['average']:.1f}%")
        if val_acc['average'] > best_val_acc:
            best_val_acc = val_acc['average']
            torch.save({'model_state_dict': model.state_dict(), 'val_acc': best_val_acc,
                        'category_encoder': le_category, 'color_encoder': le_color,
                        'season_encoder': le_season, 'usage_encoder': le_usage}, model_save_path)
            print(f"  ✅ Saved ({best_val_acc:.2f}%)")
    return best_val_acc

# ── Run experiments (comment out if you just want the class definition) ──
for exp_id, exp_config in EXPERIMENTS.items():
    run_experiment(exp_config, train_df, val_df, num_epochs=15)

## Florence and Integration

### Load Florence into memory

In [6]:
 # Cell 7 — Install Florence-2 dependencies:

!pip install -q transformers==4.44.2 timm einops
!pip install flash_attn --no-build-isolation -q
print("Florence-2 dependencies installed.")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 83.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 32.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 107.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.4/8.4 MB 113.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for flash_attn
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (flash_attn)
Florence-2 dependencies installed.


In [7]:
 # Cell 8 — Load Florence-2 model:

import sys
import types
import importlib.util

flash_attn_mock = types.ModuleType('flash_attn')
flash_attn_mock.__spec__ = importlib.util.spec_from_loader('flash_attn', loader=None)
sys.modules['flash_attn'] = flash_attn_mock
sys.modules['flash_attn.bert_padding'] = types.ModuleType('flash_attn.bert_padding')
sys.modules['flash_attn.flash_attn_interface'] = types.ModuleType('flash_attn.flash_attn_interface')

import torch
from transformers import AutoProcessor, AutoModelForCausalLM

DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
DTYPE = torch.float16 if torch.cuda.is_available() else torch.float32

print(f"Loading Florence-2 on {DEVICE}...")
florence_processor = AutoProcessor.from_pretrained("microsoft/florence-2-base", trust_remote_code=True)
florence_model = AutoModelForCausalLM.from_pretrained(
    "microsoft/florence-2-base",
    torch_dtype=DTYPE,
    trust_remote_code=True,
    _attn_implementation="eager"
).to(DEVICE)
print("Florence-2 ready.")

Loading Florence-2 on cuda...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


preprocessor_config.json:   0%|          | 0.00/806 [00:00<?, ?B/s]

processing_florence2.py: 0.00B [00:00, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/microsoft/florence-2-base:
- processing_florence2.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


tokenizer_config.json:   0%|          | 0.00/34.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

config.json: 0.00B [00:00, ?B/s]

configuration_florence2.py: 0.00B [00:00, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/microsoft/florence-2-base:
- configuration_florence2.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.
/usr/local/lib/python3.12/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


modeling_florence2.py: 0.00B [00:00, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/microsoft/florence-2-base:
- modeling_florence2.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


model.safetensors:   0%|          | 0.00/463M [00:00<?, ?B/s]

Florence-2 ready.


### Look at 2000 images and make descriptions for them

In [8]:
import subprocess
result = subprocess.run(['find', '/content', '-name', 'styles.csv'], capture_output=True, text=True)
print(result.stdout)

/content/drive/MyDrive/closetgenius1/Data1/fashion-dataset/styles.csv
/content/drive/MyDrive/ClosetGeniusCNN/styles.csv



In [9]:
# Cell A — Florence-2 scans training images, saves labels to JSON:

import json, os
from tqdm import tqdm
import pandas as pd
from PIL import Image

CACHE_PATH = "/content/florence2_labels.json"
MAX_IMAGES = 200  # increase if you have more Colab GPU time

# Load dataset CSV
styles_df = pd.read_csv("/content/drive/MyDrive/ClosetGeniusCNN/styles.csv", on_bad_lines='skip')
styles_df['image_path'] = styles_df['id'].astype(str).apply(lambda x: f"/content/drive/MyDrive/ClosetGeniusCNN/images/{x}.jpg")
styles_df = styles_df[styles_df['image_path'].apply(os.path.exists)].head(MAX_IMAGES)

# Load cache if exists
florence_labels = json.load(open(CACHE_PATH)) if os.path.exists(CACHE_PATH) else {}
print(f"Processing {len(styles_df)} images ({len(florence_labels)} already cached)...")

def run_task(image, task):
    inputs = florence_processor(text=task, images=image, return_tensors="pt").to(DEVICE, DTYPE)
    with torch.no_grad():
        ids = florence_model.generate(
            input_ids=inputs["input_ids"],
            pixel_values=inputs["pixel_values"],
            max_new_tokens=256, num_beams=1,
        )
    raw = florence_processor.batch_decode(ids, skip_special_tokens=False)[0]
    return florence_processor.post_process_generation(raw, task=task, image_size=image.size).get(task, "")

new_count = 0
for _, row in tqdm(styles_df.iterrows(), total=len(styles_df)):
    img_id = str(int(row['id']))
    if img_id in florence_labels:
        continue
    try:
        image = Image.open(row['image_path']).convert("RGB")
        caption = run_task(image, "<DETAILED_CAPTION>")
        tags_raw = run_task(image, "<GENERATE_TAGS>")
        tags = [t.strip() for t in str(tags_raw).split(",") if t.strip()]
        florence_labels[img_id] = {
            "caption": caption, "tags": tags,
            "orig_category": str(row.get('masterCategory', '')).lower(),
            "orig_color":    str(row.get('baseColour', '')).lower(),
            "orig_season":   str(row.get('season', '')).lower(),
            "orig_usage":    str(row.get('usage', '')).lower(),
        }
        new_count += 1
        if new_count % 100 == 0:
            json.dump(florence_labels, open(CACHE_PATH, 'w'))
            print(f"Checkpoint: {len(florence_labels)} saved")
    except:
        pass

json.dump(florence_labels, open(CACHE_PATH, 'w'))
print(f"Done. {new_count} new labels, {len(florence_labels)} total.")


Processing 200 images (0 already cached)...


  0%|          | 0/200 [00:00<?, ?it/s]/usr/local/lib/python3.12/dist-packages/transformers/generation/configuration_utils.py:615: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(
 50%|█████     | 100/200 [02:39<01:27,  1.15it/s]

Checkpoint: 100 saved


100%|██████████| 200/200 [04:03<00:00,  1.22s/it]

Checkpoint: 200 saved
Done. 200 new labels, 200 total.


### Convert decriptions into structured labels

In [10]:
# Cell B — Extract structured labels from Florence-2 output (no API calls, runs instantly):

CATEGORY_KEYWORDS = {
    'tops':       ['shirt','top','blouse','tshirt','t-shirt','sweater','hoodie','tank','polo','pullover','sweatshirt','crop'],
    'bottoms':    ['jeans','pants','trousers','shorts','skirt','leggings','chinos','joggers'],
    'dresses':    ['dress','gown','romper','jumpsuit','frock'],
    'outerwear':  ['jacket','coat','blazer','cardigan','vest','parka','windbreaker'],
    'shoes':      ['shoes','sneakers','boots','heels','sandals','loafers','flats','footwear'],
    'accessories':['bag','belt','hat','scarf','wallet','watch','cap','purse','backpack','sunglasses'],
}
COLOR_KEYWORDS = {
    'black':    ['black','ebony'], 'white': ['white','ivory','off-white'],
    'gray':     ['gray','grey','charcoal','silver'], 'navy': ['navy','midnight blue'],
    'blue':     ['blue','cobalt','sapphire'], 'red': ['red','crimson','scarlet'],
    'green':    ['green','olive','emerald','sage','khaki'], 'brown': ['brown','tan','camel','chocolate'],
    'pink':     ['pink','rose','blush','coral','salmon'], 'purple': ['purple','violet','lavender','lilac'],
    'yellow':   ['yellow','mustard','gold'], 'orange': ['orange','rust','terracotta'],
    'burgundy': ['burgundy','wine'], 'teal': ['teal','turquoise','aqua'],
    'cream':    ['cream','beige','ecru'],
}
SEASON_KEYWORDS = {
    'summer':    ['summer','lightweight','linen','tank','sundress','breathable'],
    'winter':    ['winter','wool','coat','fleece','thermal','heavy','warm'],
    'spring':    ['spring','floral','trench','light jacket'],
    'fall':      ['fall','autumn','knit','layered'],
    'allSeason': ['versatile','classic','denim','year round','all season'],
}
USAGE_KEYWORDS = {
    'casual':  ['casual','everyday','relaxed','streetwear','jeans','sneakers'],
    'formal':  ['formal','suit','blazer','professional','business','elegant'],
    'sporty':  ['sport','athletic','gym','workout','running','activewear'],
    'ethnic':  ['ethnic','traditional','embroidered','cultural'],
}

def best_match(text, keyword_map, fallback):
    text = text.lower()
    scores = {label: sum(1 for kw in kws if kw in text) for label, kws in keyword_map.items()}
    best = max(scores, key=scores.get)
    return best if scores[best] > 0 else fallback

refined_labels = {}
for img_id, item in florence_labels.items():
    blob = " ".join([item.get("caption", "")] + item.get("tags", []))
    refined_labels[img_id] = {
        **item,
        "category": best_match(blob, CATEGORY_KEYWORDS, item.get("orig_category", "tops")),
        "color":    best_match(blob, COLOR_KEYWORDS,    item.get("orig_color",    "black")),
        "season":   best_match(blob, SEASON_KEYWORDS,   item.get("orig_season",   "allSeason")),
        "usage":    best_match(blob, USAGE_KEYWORDS,    item.get("orig_usage",    "casual")),
    }

json.dump(refined_labels, open("/content/refined_labels.json", 'w'))

# Show a sample comparison
s = list(refined_labels.values())[0]
print(f"Caption: {s['caption'][:120]}")
print(f"Original → category:{s['orig_category']}  color:{s['orig_color']}  season:{s['orig_season']}")
print(f"Refined  → category:{s['category']}  color:{s['color']}  season:{s['season']}")
print(f"\nTotal refined: {len(refined_labels)}")



Caption: The image shows a pair of adidas originals rivalry low sneakers for women. They are a classic style with a white leather
Original → category:footwear  color:white  season:fall
Refined  → category:shoes  color:black  season:allSeason

Total refined: 200


### Retrain alexnet using these imrpoved labels

In [13]:
import torch, torch.nn as nn, torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from torchvision import transforms
from sklearn.preprocessing import LabelEncoder
from PIL import Image
import pandas as pd
import json, os

with open("/content/refined_labels.json") as f:
    refined = json.load(f)

# Merge with full dataset
styles_df = pd.read_csv("/content/drive/MyDrive/ClosetGeniusCNN/styles.csv", on_bad_lines='skip')
styles_df['image_path'] = styles_df['id'].astype(str).apply(lambda x: f"/content/drive/MyDrive/ClosetGeniusCNN/images/{x}.jpg")
styles_df = styles_df[styles_df['image_path'].apply(os.path.exists)]

merged = {}
for _, row in styles_df.iterrows():
    img_id = str(int(row['id']))
    if img_id in refined:
        merged[img_id] = refined[img_id]
    else:
        merged[img_id] = {
            "category": str(row.get('masterCategory', 'tops')).lower(),
            "color":    str(row.get('baseColour', 'black')).lower(),
            "season":   str(row.get('season', 'allSeason')).lower(),
            "usage":    str(row.get('usage', 'casual')).lower(),
        }

refined = merged
print(f"Retraining on {len(refined)} samples")

class RefinedDataset(Dataset):
    def __init__(self, data, transform=None):
        self.items = list(data.items())
        self.transform = transform
        self.cat_enc = LabelEncoder().fit([d['category'] for _, d in self.items])
        self.col_enc = LabelEncoder().fit([d['color']    for _, d in self.items])
        self.sea_enc = LabelEncoder().fit([d['season']   for _, d in self.items])
        self.use_enc = LabelEncoder().fit([d['usage']    for _, d in self.items])

    def __len__(self): return len(self.items)

    def __getitem__(self, idx):
        img_id, data = self.items[idx]
        try:
            img = Image.open(f"/content/drive/MyDrive/ClosetGeniusCNN/images/{img_id}.jpg").convert("RGB")
        except:
            img = Image.new("RGB", (224, 224), (128, 128, 128))
        if self.transform: img = self.transform(img)
        return img, {
            "category": torch.tensor(self.cat_enc.transform([data['category']])[0], dtype=torch.long),
            "color":    torch.tensor(self.col_enc.transform([data['color']])[0],    dtype=torch.long),
            "season":   torch.tensor(self.sea_enc.transform([data['season']])[0],   dtype=torch.long),
            "usage":    torch.tensor(self.use_enc.transform([data['usage']])[0],    dtype=torch.long),
        }

transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.RandomHorizontalFlip(),
    transforms.ColorJitter(0.2, 0.2, 0.2),
    transforms.ToTensor(),
    transforms.Normalize([0.485,0.456,0.406],[0.229,0.224,0.225])
])

dataset = RefinedDataset(refined, transform)
train_n = int(0.85 * len(dataset))
train_set, val_set = torch.utils.data.random_split(dataset, [train_n, len(dataset)-train_n])
train_loader = DataLoader(train_set, batch_size=32, shuffle=True, num_workers=2)
val_loader   = DataLoader(val_set,   batch_size=32, shuffle=False, num_workers=2)

print(f"Train: {train_n}  Val: {len(dataset)-train_n}")
print(f"Category classes: {dataset.cat_enc.classes_.tolist()}")

model = MultiOutputAlexNet(
    num_categories=len(dataset.cat_enc.classes_),
    num_colors=len(dataset.col_enc.classes_),
    num_seasons=len(dataset.sea_enc.classes_),
    num_usage=len(dataset.use_enc.classes_),
).to(DEVICE)

optimizer = optim.Adam(model.parameters(), lr=1e-4, weight_decay=1e-4)
scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=10)
criterion = nn.CrossEntropyLoss()
best_acc  = 0.0

for epoch in range(10):
    model.train()
    total_loss = 0
    for imgs, labels in train_loader:
        imgs = imgs.to(DEVICE)
        c, co, s, u = [labels[k].to(DEVICE) for k in ['category','color','season','usage']]
        optimizer.zero_grad()
        o_c, o_co, o_s, o_u = model(imgs)
        loss = criterion(o_c,c) + criterion(o_co,co) + criterion(o_s,s) + criterion(o_u,u)
        loss.backward(); optimizer.step()
        total_loss += loss.item()

    model.eval()
    correct, total = {k:0 for k in ['cat','col','sea','use']}, 0
    with torch.no_grad():
        for imgs, labels in val_loader:
            imgs = imgs.to(DEVICE)
            o_c, o_co, o_s, o_u = model(imgs)
            correct['cat'] += (o_c.argmax(1)  == labels['category'].to(DEVICE)).sum().item()
            correct['col'] += (o_co.argmax(1) == labels['color'].to(DEVICE)).sum().item()
            correct['sea'] += (o_s.argmax(1)  == labels['season'].to(DEVICE)).sum().item()
            correct['use'] += (o_u.argmax(1)  == labels['usage'].to(DEVICE)).sum().item()
            total += imgs.size(0)

    avg = sum(correct.values()) / (4 * total)
    scheduler.step()
    print(f"Epoch {epoch+1}/10 | Loss:{total_loss/len(train_loader):.3f} | "
          f"Cat:{correct['cat']/total:.1%} Col:{correct['col']/total:.1%} "
          f"Sea:{correct['sea']/total:.1%} Use:{correct['use']/total:.1%} | Avg:{avg:.1%}")

    if avg > best_acc:
        best_acc = avg
        torch.save({
            'model_state_dict': model.state_dict(),
            'cat_classes': dataset.cat_enc.classes_.tolist(),
            'col_classes': dataset.col_enc.classes_.tolist(),
            'sea_classes': dataset.sea_enc.classes_.tolist(),
            'use_classes': dataset.use_enc.classes_.tolist(),
        }, '/content/drive/MyDrive/ClosetGeniusCNN/models/alexnet_refined.pth')
        print(f"  ✓ Best model saved ({best_acc:.1%})")

print(f"\nDone. Best avg accuracy: {best_acc:.1%}")

Retraining on 6998 samples
Train: 5948  Val: 1050
Category classes: ['accessories', 'apparel', 'dresses', 'footwear', 'free items', 'outerwear', 'shoes', 'tops']
Epoch 1/10 | Loss:3.481 | Cat:97.8% Col:61.2% Sea:64.4% Use:89.5% | Avg:78.2%
  ✓ Best model saved (78.2%)
Epoch 2/10 | Loss:2.512 | Cat:98.0% Col:60.1% Sea:66.3% Use:88.6% | Avg:78.2%
Epoch 3/10 | Loss:2.131 | Cat:98.1% Col:63.7% Sea:71.0% Use:90.5% | Avg:80.8%
  ✓ Best model saved (80.8%)
Epoch 4/10 | Loss:1.824 | Cat:97.9% Col:64.0% Sea:68.0% Use:90.9% | Avg:80.2%
Epoch 5/10 | Loss:1.584 | Cat:97.7% Col:65.2% Sea:71.5% Use:91.4% | Avg:81.5%
  ✓ Best model saved (81.5%)
Epoch 6/10 | Loss:1.343 | Cat:97.3% Col:64.7% Sea:71.4% Use:91.0% | Avg:81.1%
Epoch 7/10 | Loss:1.171 | Cat:97.8% Col:65.5% Sea:72.5% Use:91.0% | Avg:81.7%
  ✓ Best model saved (81.7%)
Epoch 8/10 | Loss:1.008 | Cat:97.9% Col:65.2% Sea:73.7% Use:91.3% | Avg:82.0%
  ✓ Best model saved (82.0%)
Epoch 9/10 | Loss:0.916 | Cat:98.0% Col:66.1% Sea:73.3% Use:91.0% | A

##integration into app
 - Ngrok is a tunneling tool that creates a temporary public URL that forwards traffic to my Flask server running inside Google Colab. Since the AlexNet model is too large to run directly on an iPhone, I needed a way for my Swift app to send photos to a server that runs the model and returns the predictions. Without ngrok, the Flask server would only be accessible inside Google's infrastructure with no way for the iPhone to reach it. I chose ngrok specifically because it lets me iterate quickly during development without needing to set up or pay for cloud infrastructure like AWS or Google Cloud. It's a development solution — down the line I'd migrate the Flask server to a permanent cloud host and replace the ngrok URL with a stable one, but for now it gets the job done and keeps the setup simple while I'm still building and improving the model.

In [ ]:
# ===================================
# CELL 4: FLASK INFERENCE API
# ===================================

!pip install -q flask pyngrok

import torch
import torch.nn as nn
from torchvision import transforms, models
from PIL import Image
import io
import os
from flask import Flask, request, jsonify
from pyngrok import ngrok

# ===================================
# LOAD MODEL
# ===================================

DATASET_PATH = "/content/drive/MyDrive/ClosetGeniusCNN/"
MODEL_PATH = os.path.join(DATASET_PATH, "models/Frozen_Baseline_best.pth")

device = torch.device("cpu")
print(f"Using device: {device}")

# Rebuild model architecture
class MultiOutputAlexNet(nn.Module):
    def __init__(self, num_categories, num_colors, num_seasons, num_usage):
        super(MultiOutputAlexNet, self).__init__()
        alexnet       = models.alexnet(weights=models.AlexNet_Weights.IMAGENET1K_V1)
        self.features = alexnet.features
        self.avgpool  = alexnet.avgpool
        self.shared   = nn.Sequential(
            nn.Dropout(0.5),
            nn.Linear(256 * 6 * 6, 4096),
            nn.ReLU(inplace=True),
            nn.Dropout(0.5),
            nn.Linear(4096, 2048),
            nn.ReLU(inplace=True)
        )
        self.category_head = nn.Linear(2048, num_categories)
        self.color_head    = nn.Linear(2048, num_colors)
        self.season_head   = nn.Linear(2048, num_seasons)
        self.usage_head    = nn.Linear(2048, num_usage)

    def forward(self, x):
        x = self.features(x)
        x = self.avgpool(x)
        x = torch.flatten(x, 1)
        x = self.shared(x)
        return (
            self.category_head(x),
            self.color_head(x),
            self.season_head(x),
            self.usage_head(x)
        )

# Load checkpoint
print("Loading model...")
from sklearn.preprocessing import LabelEncoder
torch.serialization.add_safe_globals([LabelEncoder])
checkpoint = torch.load(MODEL_PATH, map_location=device, weights_only=False)
le_category = checkpoint['category_encoder']
le_color    = checkpoint['color_encoder']
le_season   = checkpoint['season_encoder']
le_usage    = checkpoint['usage_encoder']

model = MultiOutputAlexNet(
    num_categories=len(le_category.classes_),
    num_colors=len(le_color.classes_),
    num_seasons=len(le_season.classes_),
    num_usage=len(le_usage.classes_)
).to(device)

model.load_state_dict(checkpoint['model_state_dict'])
model.eval()
print("✅ Model loaded successfully!")
print(f"   Categories : {list(le_category.classes_)}")
print(f"   Colors     : {list(le_color.classes_)}")
print(f"   Seasons    : {list(le_season.classes_)}")
print(f"   Usage      : {list(le_usage.classes_)}")

# ===================================
# IMAGE TRANSFORM
# ===================================

transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406],
                         [0.229, 0.224, 0.225])
])

# ===================================
# PREDICTION FUNCTION
# ===================================

def predict(image_bytes):
    image  = Image.open(io.BytesIO(image_bytes)).convert('RGB')
    tensor = transform(image).unsqueeze(0).to(device)

    with torch.no_grad():
        cat_out, col_out, sea_out, usa_out = model(tensor)

    def get_pred(output, encoder):
        probs      = torch.softmax(output, dim=1)
        confidence = probs.max().item()
        label_idx  = output.argmax(1).item()
        label      = encoder.classes_[label_idx]
        return label, round(confidence * 100, 1)

    category, cat_conf = get_pred(cat_out, le_category)
    color,    col_conf = get_pred(col_out, le_color)
    season,   sea_conf = get_pred(sea_out, le_season)
    usage,    usa_conf = get_pred(usa_out, le_usage)

    return {
        "category":            category,
        "category_confidence": cat_conf,
        "color":               color,
        "color_confidence":    col_conf,
        "season":              season,
        "season_confidence":   sea_conf,
        "usage":               usage,
        "usage_confidence":    usa_conf,
    }

# ===================================
# FLASK APP
# ===================================

app = Flask(__name__)

@app.route('/predict', methods=['POST'])
def predict_endpoint():
    if 'image' not in request.files:
        return jsonify({'error': 'No image provided'}), 400
    try:
        image_bytes = request.files['image'].read()
        result      = predict(image_bytes)
        print(f"📸 Prediction: {result}")
        return jsonify(result)
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@app.route('/health', methods=['GET'])
def health():
    return jsonify({'status': 'ok', 'model': 'Large_Batch_best'})

# ===================================
# START SERVER
# ===================================

# Paste your ngrok token here
NGROK_TOKEN = "3B3H31JuVlU1FMMqAPjN0OrTHOO_6ab3tvH9guDrarNGpGHfA"
ngrok.set_auth_token(NGROK_TOKEN)

# Kill any existing ngrok tunnels
ngrok.kill()

# Start tunnel
public_url = ngrok.connect(5000)
print("API is live!")
print("Public URL: " + str(public_url))
print("Endpoint: " + str(public_url) + "/predict")
print("Paste this URL into your Swift app as baseURL")

# Start Flask (this line blocks — cell stays running while server is live)
app.run(port=5000)

## **Trials recap:**



### First few trials


- Run 1 — 75.78% (1.5 hrs, GPU)
Original setup. Basic augmentation, 7k images, batch size 64, lr 0.001, 15 epochs. This was the best result, but need to improve the acuracy
- Run 2 — 70% (4 hrs, CPU)
Added heavy augmentation simulating real phone conditions (blur, perspective, color jitter). Accuracy dropped and took longer because I ran on CPU instead of GPU.
- Run 3 — 75% (4 hrs, CPU)
Doubled the dataset from 7k to 14k images hoping more data would help. Accuracy stayed the same as Run 1 but took way longer again because of CPU.
- Conclusion:
More data and heavier augmentation didn't beat the original simple setup. Run 1 remains the best model. The main variable that hurt Runs 2 and 3 was training on CPU instead of GPU which made them 2-3x slower for no accuracy benefit. Going back to Run 1 settings on GPU is the move.

- Some future moves: freezing early layers that have been trained well already, or oversample the smaller categories

### Trial 4

- Accuracy plateaued around 75% across multiple runs so the next step is freezing the early AlexNet layers to preserve the pretrained ImageNet features and only fine-tune the later layers. Narrowed experiments down to just Baseline and High LR since the other configs proved ineffective in previous runs."

- After testing the updated model in the app, the accuracy improved noticeably. Using a photo of black jean shorts the model correctly identified the category, color, and usage. The key improvement that got us from 75% to 78.20% was freezing the early convolutional layers of AlexNet, preserving the pretrained ImageNet features and only fine-tuning the layers relevant to clothing classification.

## Summary of events

### ✅ **1. Load and Preprocess Dataset**
- Loads CSV with error handling
- Verifies image existence
- Cleans data (removes NaN)
- Prints dataset statistics

### ✅ **2. Resize Images to 224×224**
- All images resized to AlexNet's required input size
- Proper normalization with ImageNet mean/std

### ✅ **3. Train and Validate**
- Proper train/validation split (80/20)
- Separate train and validation functions
- Progress tracking every epoch

### ✅ **4. Run Experiments**
- **4 experiments** testing:
  - Different batch sizes (32 vs 64)
  - Different learning rates (0.001 vs 0.01)
  - Different augmentation (basic vs heavy)
- Automatically compares all experiments

### ✅ **5. Log and Plot Metrics**
- **Detailed plots** for each experiment (6 charts per experiment)
- **Comparison bar chart** showing all experiments
- **JSON logs** with complete training history
- All saved to Google Drive

### ✅ **6. Save Model Weights**
- Saves **best model** for each experiment
- Includes label encoders for deployment
- Checkpoint includes optimizer state for resuming

---

## **What You'll Get:**

/MyDrive/ClosetGeniusCNN/
├── models/
│   ├── Baseline_best.pth
│   ├── Large_Batch_best.pth
│   ├── High_Learning_Rate_best.pth
│   └── Heavy_Augmentation_best.pth
├── logs/
│   ├── Baseline_log.json
│   ├── Large_Batch_log.json
│   ├── High_Learning_Rate_log.json
│   └── Heavy_Augmentation_log.json
└── plots/
    ├── Baseline_metrics.png
    ├── Large_Batch_metrics.png
    ├── High_Learning_Rate_metrics.png
    ├── Heavy_Augmentation_metrics.png
    └── experiment_comparison.png